### PR TITLE
SmallMatrix: Matrix class with compile time size

### PR DESCRIPTION
**Pull request by WeiqunZhang**
Add amrex::SmallMatrix class with compile time size.

Useful for, e.g.:
- https://github.com/ECP-WarpX/impactx/pull/714
- https://github.com/ECP-WarpX/impactx/pull/702

### CODE REVIEW

**Comment by zingale** [comment]
it would be good to have a `.setVal()` method to set all elements to a value.

**Comment by WeiqunZhang** [comment]
Notes on why SmallMatrix matrix{} is zero initialized.

SmallMatrix is not an aggregate, because it has a user declared default
constructor. The rule is that, for `SmallMatrix matrix{}` with an empty
brace-enclosed initializer list, value-initialization is performed. The
effects of value-initialization of SmallMatrix (which has a user-declared
but not user-provided default constructor) are that the matrix object is
first zero-initialized and then the object's default constructor is
applied. Since the default constructor does nothing, the final result is
the object is zero-initialized.

Why is SmallMatrix's default constructor user-declared not user-provided?
It's because we first declare it with `SmallMatrix () = default`.

Reference:
  https://en.cppreference.com/w/cpp/language/list_initialization
  https://en.cppreference.com/w/cpp/language/value_initialization
  https://en.cppreference.com/w/cpp/language/zero_initialization

**Comment by ax3l** [comment]
Is `SmallMatrix<Real, N, M> m;` also zero-defined then or is it allocated but values are undefined?

**Comment by ax3l** [comment]
FYI @cemitch99

File: Tests/SmallMatrix/main.cpp [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+                AMREX_ALWAYS_ASSERT(m34(i,j) == v++);
+            }
+        }
+    }
+    {
+        auto v3 = SmallVector<double,3>::Zero();
+        v3[0] = 1.;
+        v3(1) = 2.;
+        v3[2] = 3.;
+        auto m33 = SmallMatrix<double,3,3>::Identity();
```
Comment (ax3l) [inline Tests/SmallMatrix/main.cpp:66]: Nice, very useful!

**Comment by WeiqunZhang** [comment]
It's uninitialized by default.
```
/**
 * \brief Default constructor
 *
 * The data are uninitialized by default. If you want to initialize
 * to zero, you can do `SmallMatrix<T,NRows,NCols> M{};`.
 */
AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
constexpr SmallMatrix () = default;
```

**Review by ax3l — APPROVED** [review]
That is super useful, I think that is all we currently need :rocket: :sparkles:

**Pull request merged**